### PR TITLE
Implement strncmp

### LINKED ISSUE #459: Substring/partial string comparison
> Yes, this is super annoying so a fix would be good to see soon. We do need more than this though for pattern matching in a predicate. I  often need to do the equivalent of 't:syscall:whatever/comm == "hhvm*" as I am interested in a range of processes whose name begins with "hhvm". In DTrace you could do something like /strstr(execname, "hhvm") != NULL/ if my memory serves me right.

_Originally posted by @tyroguru in https://github.com/iovisor/bpftrace/issues/198#issuecomment-461229979_

### PR DESCRIPTION
**Pull request by jgkamat**
I'll be honest and say I mostly have no idea what I'm doing here, so I'd appreciate it if I could get some extra review and guidance for this. This is my first time working with any serious compiler stuff, and my first real exposure to LLVM. I know you're probably busy for the 0.9.1 release, so feel free to put this on the backburner until you have more time.

For some reason:

```
BEGIN { if (strncmp("hhvm", str($1), 4)) { printf("hello!"); } }
```
results in 
```
bpftrace: /usr/include/llvm/IR/Instructions.h:1117: void llvm::ICmpInst::AssertOK(): Assertion `getOperand(0)->getType() == getOperand(1)->getType() && "Both operands to ICmp instruction are not of the same type!"' failed.
```
but, `if(1)` or `if(strncmp("hhvm", str($1), -1) != 0)` work fine. I think I'm missing something in the semantic analyzer but I'm not sure. It might need some changes to the `if` logic, but I'm not sure what/where those would be.

Fixes #459

### CODE REVIEW

File: src/ast/codegen_llvm.cpp [hunk 1]
```diff
@@ -758,7 +758,26 @@ void CodegenLLVM::visit(Call &call)
     }
     expr_ = stackid;
   }
-
+  else if (call.func == "strncmp") {
+    uint64_t size = static_cast<Integer *>(call.vargs->at(2))->n;
+    const auto& left_arg = call.vargs->at(0);
+    const auto& right_arg = call.vargs->at(1);
+
+    if (right_arg->is_literal) {
```
Comment (jgkamat) [inline src/ast/codegen_llvm.cpp]: Is there a reason these literal checks are required for the `==` string check? This version at least seems to work without them. Is it to improve performance on an inline string match?

File: tests/codegen/strncmp.cpp [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+namespace bpftrace {
+namespace test {
+namespace codegen {
+
+TEST(codegen, strncmp_test)
+{
+  test("kretprobe:vfs_read /strncmp(comm, \"sshd\", 2)/ { @[comm] = count(); }",
+
+R"EXPECTED(; Function Attrs: nounwind
```
Comment (jgkamat) [inline tests/codegen/strncmp.cpp]: Do I need to have a LLVM < 6 block for this tests (the one I was emulating had it). I don't have a very easy access to older llvms, but I can try and get it if needed.

File: src/ast/codegen_llvm.cpp (hunk 1 above)
Comment (mmarchini) [reply src/ast/codegen_llvm.cpp]: Yeah, this is an optimization so we don't have to save the literal string on memory just to load it later. `==` would work without it as well. I suggest keeping it.

File: tests/codegen.cpp [hunk 3]
```diff
@@ -77,6 +77,7 @@
 #include "codegen/pred_binop.cpp"
 #include "codegen/string_equal_comparison.cpp"
 #include "codegen/string_not_equal_comparison.cpp"
+#include "codegen/strncmp.cpp"
```
Comment (mmarchini) [inline tests/codegen.cpp]: This is not needed anymore :)

**Comment by mmarchini** [comment]
Overall code lgtm. Will do some more testing later.

---

> For some reason:
>
> ``` 
> BEGIN { if (strncmp("hhvm", str($1), 4)) { printf("hello!"); } }
> ```
>
> results in
> 
> ```
> bpftrace: /usr/include/llvm/IR/Instructions.h:1117: void llvm::ICmpInst::AssertOK(): Assertion `getOperand(0)->getType() == getOperand(1)->getType() && "Both operands to ICmp instruction are not of the same type!"' failed.
> ```
> 
> but, if(1) or if(strncmp("hhvm", str($1), -1) != 0) work fine.

LLVM is pretty strict with Int types, it won't let you use most binary operators between two integers of different size. `CreateStrncmp` will save the result of the comparison as an Int8 ([irbuilderbpf.cpp#L376](https://github.com/iovisor/bpftrace/blob/master/src/ast/irbuilderbpf.cpp#L376), [irbuilderbpf.cpp#L430](https://github.com/iovisor/bpftrace/blob/master/src/ast/irbuilderbpf.cpp#L430)), and `if` will try to compare it to a `Int64` (which is always zero, [codegen_llvm.cpp#L1127](https://github.com/iovisor/bpftrace/blob/master/src/ast/codegen_llvm.cpp#L1127)). So why `==` works? Turns out we uppercast our Binop results to Int64 to try to avoid this issue ([codegen_llvm.cpp#L796](https://github.com/iovisor/bpftrace/blob/master/src/ast/codegen_llvm.cpp#L796)). Changing the type used to store the result strcmp to Int64 fixes the issue:

```diff
diff --git a/src/ast/irbuilderbpf.cpp b/src/ast/irbuilderbpf.cpp
index c8aa040..9778aed 100644
--- a/src/ast/irbuilderbpf.cpp
+++ b/src/ast/irbuilderbpf.cpp
@@ -378,7 +378,7 @@ Value *IRBuilderBPF::CreateStrcmp(Value* val, std::string str, bool inverse) {
 Value *IRBuilderBPF::CreateStrncmp(Value* val, std::string str, uint64_t n, bool inverse) {
   Function *parent = GetInsertBlock()->getParent();
   BasicBlock *str_ne = BasicBlock::Create(module_.getContext(), "strcmp.false", parent);
-  AllocaInst *store = CreateAllocaBPF(getInt8Ty(), "strcmp.result");
+  AllocaInst *store = CreateAllocaBPF(getInt64Ty(), "strcmp.result");
 
   CreateStore(getInt1(inverse), store);
 
@@ -436,7 +436,7 @@ Value *IRBuilderBPF::CreateStrncmp(Value* val1, Value* val2, uint64_t n, bool in
   */
   Function *parent = GetInsertBlock()->getParent();
   BasicBlock *str_ne = BasicBlock::Create(module_.getContext(), "strcmp.false", parent);
-  AllocaInst *store = CreateAllocaBPF(getInt8Ty(), "strcmp.result");
+  AllocaInst *store = CreateAllocaBPF(getInt64Ty(), "strcmp.result");
   BasicBlock *done = BasicBlock::Create(module_.getContext(), "strcmp.done", parent);
 
   CreateStore(getInt1(inverse), store);
```

File: tests/codegen/strncmp.cpp (hunk 2 above)
Comment (mmarchini) [reply tests/codegen/strncmp.cpp]: I can run the tests on the LLVM versions failing on CI (I have them easily set up on my machine). Can you rebase and fix the `if` expression issue first? The code generated might change a little bit after that.

**Comment by jgkamat** [comment]
Thanks for the pointers, that was very helpful! While trying to fix this test:
```
bpftrace -e 'BEGIN { if (!strncmp(str($1), str($2), 7)) { printf("not equal");} exit();}' "python3-proc" "hhvm"
```

I discovered that `!` dosen't work well inside if statements, for example, this fails:

```
bpftrace -e 'BEGIN { $x = 5; if (!$x) { printf("not equal");} exit();}'
```

I'll ignore that for now because that seems like a seperate issue though.

File: tests/codegen/string_equal_comparison.cpp [hunk 4]
```diff
@@ -36,95 +30,51 @@ define i64 @"kretprobe:vfs_read"(i8* nocapture readnone) local_unnamed_addr sect
...
 
 strcmp.loop9:                                     ; preds = %strcmp.loop5
   call void @llvm.lifetime.start.p0i8(i64 -1, i8* nonnull %strcmp.char14)
-  %11 = add [16 x i8]* %comm, i64 4
-  %probe_read15 = call i64 inttoptr (i64 4 to i64 (i8*, i64, i8*)*)(i8* nonnull %strcmp.char14, i64 1, [16 x i8]* %11)
-  %12 = load i8, i8* %strcmp.char14, align 1
-  %strcmp.cmp16 = icmp eq i8 %12, 0
-  br i1 %strcmp.cmp16, label %pred_true, label %pred_false
-
-lookup_success:                                   ; preds = %pred_true
```
Comment (fbs) [inline tests/codegen/string_equal_comparison.cpp]: Why is this whole section dissapearing? That can't be ok

File: src/ast/irbuilderbpf.cpp [hunk 5]
```diff
@@ -427,14 +436,14 @@ Value *IRBuilderBPF::CreateStrcmp(Value* val1, Value* val2, bool inverse) {
   */
   Function *parent = GetInsertBlock()->getParent();
   BasicBlock *str_ne = BasicBlock::Create(module_.getContext(), "strcmp.false", parent);
-  AllocaInst *store = CreateAllocaBPF(getInt8Ty(), "strcmp.result");
+  AllocaInst *store = CreateAllocaBPF(getInt64Ty(), "strcmp.result");
```
Comment (fbs) [inline src/ast/irbuilderbpf.cpp]: why this change?

File: tests/codegen/string_equal_comparison.cpp (hunk 4 above)
Comment (fbs) [reply tests/codegen/string_equal_comparison.cpp]:
```
vagrant@ubuntu-cosmic:~/bpftrace$ sudo bpftrace_master -e 't:syscalls:sys_enter_openat /comm == "cat"/ { @[comm]++ }' -c "cat xxx"
Attaching 1 probe...
/bin/cat: xxx: No such file or directory


@[cat]: 25

vagrant@ubuntu-cosmic:~/bpftrace$ sudo bpftrace -e 't:syscalls:sys_enter_openat /comm == "cat"/ { @[comm]++ }' -c "cat xxx"
Attaching 1 probe...
/bin/cat: xxx: No such file or directory
```

WIth your patch the whole function body seems to be optimized away. The optimized IR doesn't contain the map lookup and store code

File: tests/runtime/other [hunk 6]
```diff
@@ -58,6 +58,16 @@ RUN bpftrace -v -e 'BEGIN { if (str($1) != str($2)) { printf("I got %s\n", str($
 EXPECT not equal
 TIMEOUT 5
 
+NAME struct partial string compare - pass
+RUN bpftrace -v -e 'BEGIN { if (strncmp(str($1), str($2), 4)) { printf("I got %s\n", str($1));} exit();}' "hhvm" "hhvm-proc"
+EXPECT I got hhvm
+TIMEOUT 5
+
+NAME struct partial string compare - pass reverse
```
Comment (fbs) [inline tests/runtime/other]: Having a more realistic test would be good imo. Something like

`bpftrace -e 't:syscalls:sys_enter_openat /comm == "cat"/ { @[comm]++ }' -c "cat xxx"`

File: src/ast/irbuilderbpf.cpp (hunk 5 above)
Comment (jgkamat) [reply src/ast/irbuilderbpf.cpp]: This was added due to [this comment](http://www.github.com/iovisor/bpftrace/pull/778#issuecomment-515167110), essentially doing a simple compare in an if statement was failing because of a type mismatch. I'm not sure why this is causing the map lookup code to be removed (reverting this does fix that), I'll try to figure out why that's happening...

File: tests/codegen/strncmp.cpp (hunk 2 above)
Comment (fbs) [reply tests/codegen/strncmp.cpp]: @mmarchini  do you have a moment to look at the LLVM5 test here?

File: src/ast/semantic_analyser.cpp [hunk 7]
```diff
@@ -559,6 +559,17 @@ void SemanticAnalyser::visit(Call &call)
   else if (call.func == "ustack") {
     check_stack_call(call, Type::ustack);
   }
+  else if (call.func == "strncmp") {
+    check_nargs(call, 3);
```
Comment (fbs) [inline src/ast/semantic_analyser.cpp]: To prevent going out of range: 

```
    if(check_nargs(call, 3)) {
      check_arg(call, Type::string, 0);
      check_arg(call, Type::string, 1);
      check_arg(call, Type::integer, 2, true);

     Integer &size = static_cast<Integer&>(*call.vargs->at(2));
     if (size.n < 0)
       err_ << "Builtin strncmp requires a non-negative size" << std::endl; 
   }
    call.type = SizedType(Type::integer, 8);
  }
```

File: tests/codegen/strncmp.cpp (hunk 2 above)
Comment (fbs) [reply tests/codegen/strncmp.cpp]: This seems to be the same issue as #186. 

It is fine with me if you add this to the excludes for the LLVM5 build

File: src/ast/semantic_analyser.cpp (hunk 7 above)
Comment (fbs) [reply src/ast/semantic_analyser.cpp]:
```
bpftrace -e 'i:s:1 { strncmp("a"); }'
terminate called after throwing an instance of 'std::out_of_range'
  what():  vector::_M_range_check: __n (which is 1) >= this->size() (which is 1)
Aborted
```

File: src/ast/semantic_analyser.cpp [hunk 8]
```diff
@@ -559,6 +559,17 @@ void SemanticAnalyser::visit(Call &call)
...
+    check_arg(call, Type::string, 0);
+    check_arg(call, Type::string, 1);
```
Comment (fbs) [inline src/ast/semantic_analyser.cpp]: Can you add a few semantic analyser tests for this? Especially the "wrong" cases are good to test, e.g.:

```
TEST(semantic_analyser, strncmp)
{
  test("i:s:1 { strncmp(1) }", 1);
  test("i:s:1 { strncmp(1,1,1) }", 1);
  test("i:s:1 { strncmp("a",1,1) }", 1);
  test("i:s:1 { strncmp("a","a",-1) }", 1);
}
```

**Comment by jgkamat** [comment]
Sure, that seems fine to me, it's a little less convenient but it is better to follow the standard strncmp return values. I don't think I can emulate the greater than/less than zero behavior without  a little more work (which I'd rather leave for a separate diff), but zero/non-zero should be easy.

File: src/ast/semantic_analyser.cpp [hunk 9]
```diff
@@ -645,6 +645,18 @@ void SemanticAnalyser::visit(Call &call)
     buf << "BPF_FUNC_send_signal not available for your kernel version";
 #endif
   }
+  else if (call.func == "strncmp") {
+    if (check_nargs(call, 3)) {
+      check_arg(call, Type::string, 0);
+      check_arg(call, Type::string, 1);
+      check_arg(call, Type::integer, 2, true);
+
+      Integer &size = static_cast<Integer&>(*call.vargs->at(2));
```
Comment (ajor) [inline src/ast/semantic_analyser.cpp]: At this point it's not guaranteed that the third argument is an integer literal, so this could lead to some dodgy memory accesses. This section needs to only run if `check_arg(call, Type::integer, 2, true)` returns true
